src/audio/asrc/asrc.c

```diff
@@ -412,6 +412,11 @@ static int asrc_params(struct comp_dev *dev,
 	/* set source/sink_frames/rate */
 	cd->source_rate = sourceb->stream.rate;
 	cd->sink_rate = sinkb->stream.rate;
+	if (cd->sink_rate == 0) {
+		comp_err(dev, "asrc_params(), zero sink rate");
```

Comment:
Do we check anywhere else for supported source /sink rates ?

Reply:
There's check in prepare() where things are set up for copy(). But this same fail would happen in prepare() if I moved this code there. I think it's better to fail early when we know things won't work.

There used to be in params() where this error happens a stream params rewrite for topology new() provided rate for sink or source but it was removed when pipeline walk was enhanced to be bi-directional.

The topology provided 48k value for sink in this is still available (ignored now) and I could revert that pipeline walk change for sink_rate == 0 or source_rate == 0 cases. In addition I should copy the PCM format from the known buffer. But I don't know if such is appropriate thing to do.

Reply:
@singalsu for my education, why would the sink_rate  be 0 for HDA or ALH? Is it because they are not using the timer-based pipeline?

Reply:
It's because in hda.c (and alh.c).

```
static int hda_get_hw_params(struct dai *dai,
			     struct sof_ipc_stream_params  *params, int dir)
{
	/* 0 means variable */
	params->rate = 0;
	params->channels = 0;
	params->buffer_fmt = 0;
	params->frame_fmt = 0;

	return 0;
}

```

Reply:
@singalsu thanks but I was wondering conceptually why it is so

Reply:
I'd assume HDA DMA can adapt to variable stream format. But I doubt it can do rate conversion or change rate per streaming session. Can you comment @tlauda @mmaka1 ?

Reply:
@singalsu it's best to check as early as possible and base the check on the supported kconfig ASRC settings and topology settings.

Reply:
The initialization is split to new(), params(), and prepare(). We've had a lot of changes in streaming start and done fixes by moving parts of initialization between these functions. As result (make small easy to review patches) it might not be the most logical any more.

But this issue is not related to Kconfig (where reject for non-supported rate happens now in prepare()).

This is related to topology such way the with the bi-directional pipeline walk the capability of SRC and ASRC to alter downstream parameters via topology provided rate was removed. We get now that information from SSP and DMIC DAI types. But not from HDA and ALH.

I could ugly fix this by implemeting to SRC components a pipeline parameters override when the stream parameter appear as zero but didn't do it because pipeline should take care of such, not individual components.

Reply:
@singalsu this makes sense but with https://github.com/thesofproject/linux/pull/1943 this problem shouldnt be seen no? But nevertheless, a check is a good thing to have.

Reply:
Yes, similar PR is needed for HDA kernel and for both the FW counterpart. Wrong parameters can and will happen in future too so it's useful to have check in component for sane parameters, especially when a crash happens.